File: 220. Contains Duplicate III/optimised.js
/**
 * @param {number[]} nums
 * @param {number} indexDiff
 * @param {number} valueDiff
 * @return {boolean}
 */
var containsNearbyAlmostDuplicate = function(nums, indexDiff, valueDiff) {
	const buckets = {};

	const bucket_size = valueDiff + 1;

	for (const [i, num] of nums.entries()) {
		const bucket_id = Math.floor(num / bucket_size);
		
		if (bucket_id in buckets) { return true; }

		if ((bucket_id - 1) in buckets && Math.abs(num - buckets[bucket_id - 1]) <= valueDiff) {
			return true;
		}

		if ((bucket_id + 1) in buckets && Math.abs(num - buckets[bucket_id + 1]) <= valueDiff) {
                        return true;
                }

		buckets[bucket_id] = num;
		if (i>=indexDiff) {
                        const old_bucket = Math.floor(nums[i - indexDiff] / bucket_size);
                        delete buckets[old_bucket];
                }

	}

	return false;
};

console.log(
    containsNearbyAlmostDuplicate([1, 2, 3, 1], 3, 0)
);

console.log(
    containsNearbyAlmostDuplicate([1, 5, 9, 1, 5, 9], 2, 3)
);

console.log(
    containsNearbyAlmostDuplicate([1, 2, 3, 1], 1, 2)
);

File: 220. Contains Duplicate III/optimised.py
#!/usr/bin/env python

from typing import List

def containsNearbyAlmostDuplicate(nums: List[int], indexDiff: int, valueDiff: int) -> bool:
    """Contains Duplicate"""
    buckets = {}
    bucket_size = valueDiff + 1

    for i, num in enumerate(nums):
        bucket_id = num // bucket_size

        if bucket_id in buckets:
            return True

        if (bucket_id - 1 in buckets and abs(num - buckets[bucket_id - 1]) <= valueDiff):
            return True

        if (bucket_id + 1 in buckets and abs(num - buckets[bucket_id + 1]) <= valueDiff):
            return True

        buckets[bucket_id] = num

        if i >= indexDiff:
            old_bucket = nums[i - indexDiff] // bucket_size
            del buckets[old_bucket]

    return False

print(containsNearbyAlmostDuplicate([1, 2, 3, 1], 3, 0))
print(containsNearbyAlmostDuplicate([1, 5, 9, 1, 5, 9], 2, 3))
